src/Compilers/Core/Portable/Compilation/SourceReferenceResolver.cs

```diff
@@ -53,5 +55,16 @@ internal Stream OpenReadChecked(string fullPath)
 
             return stream;
         }
+
+        /// <summary>
+        /// Reads the contents of <paramref name="resolvedPath"/> and returns a <see cref="SourceText"/>.
+        /// </summary>
+        /// <param name="resolvedPath">Path returned by <see cref="ResolveReference(string, string)"/>.</param>
+        /// <param name="location"><see cref="Location"/> to be used for any <see cref="Diagnostic"/> being produced.</param>
+        /// <param name="diagnostics">Any <see cref="Diagnostic"/> produced while attempting to read the contents.</param>
+        public virtual SourceText ReadText(string resolvedPath, Location location, IList<Diagnostic> diagnostics)
```

Comment:
Could the method generate a list of DiagnosticInfos rather than Diagnostics? Then the caller could set the location.